Doublecheck if we need to override properties in java.security

So far, we have few properties overriden when running with `kc.java.security` file in our testsuite. But it is possible that we don't need to override almost anything and we can still achieve everything working. This issue is about doublecheck the overriden properties in `kc.java.security` and eventually skip overriding most of the properties in this file and also make sure that people don't need to override most of the properties in their own custom manually `java.security` properties file.

Right now, the `kc.java.security` file from testsuite override those:

1) It adds `XMLDSig` security provider, which is needed for SAML. This needs to be present right now and there is separate task for it #16326 . Due to this, we would still need to have custom `kc.java.security` property files, but possibly with all the other properties removed/commented in this file with the exception of this one (See below).
 
2) The properties:
```
keystore.type=PKCS12
fips.keystore.type=PKCS12

ssl.TrustManagerFactory.algorithm=PKIX
fips.ssl.TrustManagerFactory.algorithm=PKIX
```
already have same values like the default `JAVA_HOME/conf/security/java.security` from OpenJDK17 on FIPS enabled RHEL VM. So hopefully we can just comment those properties in our `kc.java.security` ?

3) The property:
```
ssl.KeyManagerFactory.algorithm=PKIX
fips.ssl.KeyManagerFactory.algorithm=PKIX
```
has different value than default `JAVA_HOME/conf/security/java.security`. When this property is removed from the `kc.java.security` file, then some X.509 tests and authorization-adapter tests are failing (see my comment below).

So this task is mostly down to the investigation of the last property (3).

Can we just manually set this property programatically somewhere in `FIPS1402Provider` initialization with `Security.setProperty("ssl.KeyManagerFactory.algorithm", "PKIX");` ? Or would it have any bad side-effect if we do this? It will be good to doublecheck why the `X.509` tests and authorization-adapter tests are failing when this property is not overriden. Could this be related to the fact that the other party is using older algorithms for it's keys during SSL handshake? Or some completely different reason?

Thanks for all your work on this. Really, thank you!

I don't know much about Java, much less BouncyCastle.

I'm seeing this with the same BCFIPS versions as listed in pom.xml AND when creating BCFKS store types for the keystore and truststore:

```
keycloak_1  | 2023-02-06 21:27:13,399 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
keycloak_1  | 2023-02-06 21:27:13,400 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: java.lang.RuntimeException: Unable to start HTTP server
keycloak_1  |   at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:758)
keycloak_1  |   at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServer(VertxHttpRecorder.java:318)
keycloak_1  |   at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket1866188241.deploy_0(Unknown Source)
keycloak_1  |   at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket1866188241.deploy(Unknown Source)
keycloak_1  |   at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
keycloak_1  |   at io.quarkus.runtime.Application.start(Application.java:101)
keycloak_1  |   at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:110)
keycloak_1  |   at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
keycloak_1  |   at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:103)
keycloak_1  |   at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:37)
keycloak_1  |   at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
keycloak_1  |   at picocli.CommandLine.access$1300(CommandLine.java:145)
keycloak_1  |   at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
keycloak_1  |   at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
keycloak_1  |   at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
keycloak_1  |   at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
keycloak_1  |   at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
keycloak_1  |   at picocli.CommandLine.execute(CommandLine.java:2078)
keycloak_1  |   at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:93)
keycloak_1  |   at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:89)
keycloak_1  |   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
keycloak_1  |   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
keycloak_1  |   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
keycloak_1  |   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
keycloak_1  |   at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
keycloak_1  |   at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
keycloak_1  | Caused by: java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: 'int org.bouncycastle.asn1.ASN1Integer.intValueExact()'
keycloak_1  |   at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
keycloak_1  |   at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
keycloak_1  |   at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:724)
keycloak_1  |   ... 25 more
keycloak_1  | Caused by: java.lang.NoSuchMethodError: 'int org.bouncycastle.asn1.ASN1Integer.intValueExact()'
keycloak_1  |   at org.bouncycastle.asn1.cms.CCMParameters.<init>(Unknown Source)
keycloak_1  |   at org.bouncycastle.asn1.cms.CCMParameters.getInstance(Unknown Source)
keycloak_1  |   at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.decryptData(Unknown Source)
keycloak_1  |   at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.engineLoad(Unknown Source)
keycloak_1  |   at java.base/java.security.KeyStore.load(KeyStore.java:1479)
keycloak_1  |   at io.vertx.core.net.impl.KeyStoreHelper.loadKeyStore(KeyStoreHelper.java:237)
keycloak_1  |   at io.vertx.core.net.KeyStoreOptionsBase.getHelper(KeyStoreOptionsBase.java:187)
keycloak_1  |   at io.vertx.core.net.KeyStoreOptionsBase.getTrustManagerFactory(KeyStoreOptionsBase.java:217)
keycloak_1  |   at io.vertx.core.net.impl.SSLHelper.getTrustMgrFactory(SSLHelper.java:383)
keycloak_1  |   at io.vertx.core.net.impl.SSLHelper.lambda$init$0(SSLHelper.java:204)
keycloak_1  |   at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
keycloak_1  |   at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
keycloak_1  |   at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
keycloak_1  |   at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
keycloak_1  |   at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
keycloak_1  |   at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
keycloak_1  |   at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
keycloak_1  |   at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
keycloak_1  |   at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
keycloak_1  |   at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
keycloak_1  |   at java.base/java.lang.Thread.run(Thread.java:829)
keycloak_1  | 
keycloak_1 exited with code 1
```
Then I found this [comment](https://stackoverflow.com/a/74340291/7742737) that `ssl.KeyManagerFactory.algorithm=PKIX` needs to be configured.

Then I found this open issue.

I don't see the error when using PKCS12 store types.

Do you think my observed error is related to this issue?

--- a/.github/scripts/run-fips-it.sh
+++ b/.github/scripts/run-fips-it.sh
@@ -14,4 +14,5 @@ echo "STRICT_OPTIONS: $STRICT_OPTIONS"
 TESTS=`testsuite/integration-arquillian/tests/base/testsuites/suite.sh fips`
 echo "Tests: $TESTS"
 export JAVA_HOME=/etc/alternatives/java_sdk_17
+set -o pipefail
 ./mvnw test -Dsurefire.rerunFailingTestsCount=$SUREFIRE_RERUN_FAILING_COUNT -nsu -B -Pauth-server-quarkus,auth-server-fips140-2 -Dcom.redhat.fips=false $STRICT_OPTIONS -Dtest=$TESTS -pl testsuite/integration-arquillian/tests/base | misc/log/trimmer.sh

--- a/crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java
+++ b/crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java
@@ -26,15 +26,18 @@ import java.security.cert.CollectionCertStoreParameters;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 
 import javax.crypto.Cipher;
 import javax.crypto.NoSuchPaddingException;
 import javax.crypto.SecretKeyFactory;
+import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SNIHostName;
 import javax.net.ssl.SSLParameters;
 import javax.net.ssl.SSLSocket;
 import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManagerFactory;
 
 import org.bouncycastle.asn1.x9.ECNamedCurveTable;
 import org.bouncycastle.asn1.x9.X9ECParameters;
@@ -87,6 +90,8 @@ public class FIPS1402Provider implements CryptoProvider {
             checkSecureRandom(() -> Security.insertProviderAt(this.bcFipsProvider, 2));
             Provider bcJsseProvider = new BouncyCastleJsseProvider("fips:BCFIPS");
             Security.insertProviderAt(bcJsseProvider, 3);
+            // force the key and trust manager factories if default values not present in BCJSSE
+            modifyKeyTrustManagerSecurityProperties(bcJsseProvider);
             log.debugf("Inserted security providers: %s", Arrays.asList(this.bcFipsProvider.getName(),bcJsseProvider.getName()));
         } else {
             log.debugf("Security provider %s already loaded", existingBcFipsProvider.getName());
@@ -288,5 +293,43 @@ public class FIPS1402Provider implements CryptoProvider {
                 Security.setProperty("securerandom.strongAlgorithms", origStrongAlgs != null ? origStrongAlgs : "");
             }
         }
+    }
+
+    /**
+     * BCJSSE manages X.509, X509 and PKIX for KeyManagerFactory and
+     * TrustManagerFactory (names or aliases) while JSSE manages SunX509,
+     * NewSunX509 and PKIX for KeyManagerFactory and SunX509, PKIX, SunPKIX,
+     * X509 and X.509 for the TrustManagerFactory. As BCJSSE is used when
+     * fips enabled, the default implementations are changed to the ones
+     * provided by BC if selected ones are not present in the BCJSSE.
+     *
+     * @param bcJsseProvider The BCJSSE provider
+     */
+    private static void modifyKeyTrustManagerSecurityProperties(Provider bcJsseProvider) {
+        boolean setKey = bcJsseProvider.getService(KeyManagerFactory.class.getSimpleName(), KeyManagerFactory.getDefaultAlgorithm()) == null;
+        boolean setTrust = bcJsseProvider.getService(TrustManagerFactory.class.getSimpleName(), TrustManagerFactory.getDefaultAlgorithm()) == null;
+        if (!setKey && !setTrust) {
+            return;
+        }
+        Set<Provider.Service> services = bcJsseProvider.getServices();
+        if (services != null) {
+            for (Provider.Service service : services) {
+                if (setKey && KeyManagerFactory.class.getSimpleName().equals(service.getType())) {
+                    Security.setProperty("ssl.KeyManagerFactory.algorithm", service.getAlgorithm());
+                    setKey = false;
+                    if (!setTrust) {
+                        return;
+                    }
+                } else if (setTrust && TrustManagerFactory.class.getSimpleName().equals(service.getType())) {
+                    Security.setProperty("ssl.TrustManagerFactory.algorithm", service.getAlgorithm());
+                    setTrust = false;
+                    if (!setKey) {
+                        return;
+                    }
+                }
+            }
+        }
+        throw new IllegalStateException("Provider " + bcJsseProvider.getName()
+                + " does not provide KeyManagerFactory or TrustManagerFactory algorithms for TLS");
     }
 }

--- a/crypto/fips1402/src/main/java/org/keycloak/crypto/fips/KeycloakFipsSecurityProvider.java
+++ b/crypto/fips1402/src/main/java/org/keycloak/crypto/fips/KeycloakFipsSecurityProvider.java
@@ -48,7 +48,7 @@ public class KeycloakFipsSecurityProvider extends Provider {
             isSystemFipsEnabled.setAccessible(true);
             return (boolean) isSystemFipsEnabled.invoke(null);
         } catch (Throwable ignore) {
-            logger.warn("Could not detect if FIPS is enabled from the host");
+            logger.debug("Could not detect if FIPS is enabled from the host");
         } finally {
             if (isSystemFipsEnabled != null) {
                 isSystemFipsEnabled.setAccessible(false);

--- a/integration/client-cli/admin-cli/src/main/bin/kcadm.bat
+++ b/integration/client-cli/admin-cli/src/main/bin/kcadm.bat
@@ -5,4 +5,4 @@ if "%OS%" == "Windows_NT" (
 ) else (
   set DIRNAME=.\
 )
-java %KC_OPTS% -cp "%DIRNAME%\client\keycloak-admin-cli-${project.version}.jar" -Dkc.lib.dir="%DIRNAME%\client\lib" org.keycloak.client.admin.cli.KcAdmMain %*
+java %KC_OPTS% -cp "%DIRNAME%\client\keycloak-admin-cli-${project.version}.jar" --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.lib.dir="%DIRNAME%\client\lib" org.keycloak.client.admin.cli.KcAdmMain %*

--- a/integration/client-cli/admin-cli/src/main/bin/kcadm.sh
+++ b/integration/client-cli/admin-cli/src/main/bin/kcadm.sh
@@ -29,4 +29,4 @@ if [ "x$JAVA" = "x" ]; then
     fi
 fi
 
-"$JAVA" $KC_OPTS -cp $DIRNAME/client/keycloak-admin-cli-${project.version}.jar -Dkc.lib.dir=$DIRNAME/client/lib org.keycloak.client.admin.cli.KcAdmMain "$@"
+"$JAVA" $KC_OPTS -cp $DIRNAME/client/keycloak-admin-cli-${project.version}.jar --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.lib.dir=$DIRNAME/client/lib org.keycloak.client.admin.cli.KcAdmMain "$@"

--- a/integration/client-cli/client-registration-cli/src/main/bin/kcreg.bat
+++ b/integration/client-cli/client-registration-cli/src/main/bin/kcreg.bat
@@ -5,4 +5,4 @@ if "%OS%" == "Windows_NT" (
 ) else (
   set DIRNAME=.\
 )
-java %KC_OPTS% -cp "%DIRNAME%\client\keycloak-client-registration-cli-${project.version}.jar" -Dkc.lib.dir="%DIRNAME%\client\lib" org.keycloak.client.registration.cli.KcRegMain %*
+java %KC_OPTS% -cp "%DIRNAME%\client\keycloak-client-registration-cli-${project.version}.jar" --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.lib.dir="%DIRNAME%\client\lib" org.keycloak.client.registration.cli.KcRegMain %*

--- a/integration/client-cli/client-registration-cli/src/main/bin/kcreg.sh
+++ b/integration/client-cli/client-registration-cli/src/main/bin/kcreg.sh
@@ -28,4 +28,4 @@ if [ "x$JAVA" = "x" ]; then
 fi
 
 DIRNAME=`dirname "$RESOLVED_NAME"`
-"$JAVA" $KC_OPTS -cp $DIRNAME/client/keycloak-client-registration-cli-${project.version}.jar -Dkc.lib.dir=$DIRNAME/client/lib org.keycloak.client.registration.cli.KcRegMain "$@"
+"$JAVA" $KC_OPTS -cp $DIRNAME/client/keycloak-client-registration-cli-${project.version}.jar --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.lib.dir=$DIRNAME/client/lib org.keycloak.client.registration.cli.KcRegMain "$@"

--- a/testsuite/integration-arquillian/servers/auth-server/common/fips/kc.java.security
+++ b/testsuite/integration-arquillian/servers/auth-server/common/fips/kc.java.security
@@ -5,55 +5,10 @@
 # and the non-fips based (EG. when running the tests on GH actions)
 
 #
-# List of providers and their preference orders (see above). Used on the host without FIPS (EG. when running the tests on GH actions)
-# Uses only BouncyCastle FIPS providers to make sure to use only FIPS compliant cryptography.
-#
-#security.provider.1=org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider
-#security.provider.2=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
-#security.provider.3=
-
-#
 # Security providers used when global crypto-policies are set to FIPS (Usually it is used when FIPS enabled on system/JVM level)
 #
 # NOTE: This list of providers is needed to be override just because XMLDSig provider is not yet present on the OpenJDK 17 by default on the RHEL FIPS host on OpenJDK 17.0.3.
 # However once it is present, there won't be a need to override this and this part can be fully commented/removed.
 # TODO: Comment/remove this once https://bugzilla.redhat.com/show_bug.cgi?id=1940064 is fixed and OpenJDK 17 updated to corresponding version where XMLDSig is available by default
 #
-fips.provider.1=SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
-fips.provider.2=SUN
-fips.provider.3=SunEC
-fips.provider.4=SunJSSE
-fips.provider.5=SunJCE
-fips.provider.6=SunRsaSign
 fips.provider.7=XMLDSig
-
-# Commented this provider for now (and also other providers) as it uses lots of non-FIPS services.
-# See https://access.redhat.com/documentation/en-us/openjdk/11/html-single/configuring_openjdk_11_on_rhel_with_fips/index#ref_openjdk-default-fips-configuration_openjdk
-# fips.provider.2=SUN
-
-#
-# Default keystore type.
-#
-keystore.type=PKCS12
-fips.keystore.type=PKCS12
-
-# This is needed especially if we cannot add security provider "com.sun.net.ssl.internal.ssl.Provider BCFIPS" as a security provider.
-# OpenJDK has "SunX509" as default algorithm, but that one is not supported by BCJSSE. So adding the Sun provider delegating to BCFIPS is needed (as above)
-# or changing default algorithm as described here
-ssl.KeyManagerFactory.algorithm=PKIX
-fips.ssl.KeyManagerFactory.algorithm=PKIX
-
-ssl.TrustManagerFactory.algorithm=PKIX
-fips.ssl.TrustManagerFactory.algorithm=PKIX
-
-#
-# Controls compatibility mode for JKS and PKCS12 keystore types.
-#
-# When set to 'true', both JKS and PKCS12 keystore types support loading
-# keystore files in either JKS or PKCS12 format. When set to 'false' the
-# JKS keystore type supports loading only JKS keystore files and the PKCS12
-# keystore type supports loading only PKCS12 keystore files.
-#
-# This is set to true as when set to false on OpenJDK 17 and PKCS12 is default keystore type, loading of default truststore (from java cacerts) fails.
-#keystore.type.compat=false
-#fips.keystore.type.compat=false

